### ISSUE TITLE
How to get both training and test results?

### ISSUE THREAD

**Issue by tangandrew** [kaixin96/mixreg#2]
Hi, I'm trying to run your code with the following command:
`mpiexec -np 1 python -m train_procgen.train --env_name starpilot --num_levels 500 --mix_mode mixreg --test_worker_interval 1 --gpus_id 0`

However it writes to only one file, progress.csv, which looks like the training results. I tried running the scripts in the experiments directory, but I only have one GPU while those require three. How can I get both the training and the test results?

**Comment by kaixin96** [comment]
Hi @tangandrew , you may need hack some source code to enable running on a single GPU.

First, please make sure the GPU memory is large enough to run the training and testing worker processes. You can have 3 processes (2 training and 1 testing) or just 2 (1 training and 1 testing). As long as the number of total training environments is same (e.g. 256 for the current setup), the performance will be largely the same. You may use a smaller number of environments for testing, say 64, to reduce GPU memory usage. So all in all, the memory bottleneck is training with 256 parallel environments plus testing with 64 (or even fewer) parallel environments.

Second, you may need to disable `gpu_options.allow_growth` and explicitly allocate the amount of GPU memory to each process. Otherwise, TensorFlow will allocate a full GPU to one process.

If you have further questions, feel free to reply. Thank you.

Best,
Kaixin

**Comment by tangandrew** [comment]
Hi Kaixin,
Thanks for the feedback. I am now using 2 processes with 128 environments for training and 64 for testing. I also reduced the `nminibatches` to 1.  First I tried leaving `gpu_options.allow_growth = True` and then I explicitly allocated GPU memory with `config.gpu_options.per_process_gpu_memory_fraction = 0.5`, however both approaches result in Resource exhausted: OOM when allocating tensor. 

Do you have any tips on reducing GPU memory usage to 11GB? 

Thanks,
Andrew

**Comment by kaixin96** [comment]
Hi Andrew, unfortunately, 11GB is not enough to run the experiments. If I remember correctly, one process of training with 128 environments will take about 11GB GPU memory. Most of my experiments are run with 3 x 12GB Titan X. So at least 22-24 GB GPU memory is needed for training with 256 parallel environments. For testing, you can save checkpoints and evaluate it after training finishes.

Thank you,
Kaixin

**Comment by tangandrew** [comment]
Oh I see, thanks. I've been tracing the code and both the training and testing workers call `ppo2.learn`. I just want to confirm, `ppo2.learn` is in evaluate/test mode if `mpi_rank_weight` is 0 and `num_levels` is 0? 

During training, I will set `save_interval=5` so I have the model after every 5 updates. And if I want to evaluate a checkpoint after training, I can call `ppo2.learn` with `mpi_rank_weight=0`, `num_levels=0`, `load_path` set to the checkpoint path, and run one update (set `total_timesteps = nenvs * nsteps` so that `nupdates=1`)?

Thanks,
Andrew

**Comment by kaixin96** [comment]
Yes, you are right. The differences between train mode and test mode are  `mpi_rank_weight` and `num_levels `. As for testing, your solution is ok. By setting `mpi_rank_weight=0`, the gradients will not be used to update the parameters. But the training cost is still there. You can totally avoid training by only doing the rollout, i.e., collecting data with runner, though this may need some hacking.

Thank you,
Kaixin

**Comment by tangandrew** [comment]
Thanks for the help. I figured out a way to collect the test results by only doing the rollout, and avoid the training cost.
